crates/ruma-server-util/src/authorization.rs

```
@@ -37,204 +43,164 @@ impl XMatrix {
                     if origin.is_some() {
-                        debug!("Field origin duplicated in X-Matrix Authorization header");
+                        debug!("origin parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        origin = Some(OwnedServerName::try_from(value.to_unescaped())?);
                     }
-                    origin = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "destination" => {
                     if destination.is_some() {
-                        debug!("Field destination duplicated in X-Matrix Authorization header");
+                        debug!("destination parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        destination = Some(OwnedServerName::try_from(value.to_unescaped())?);
                     }
-                    destination = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "key" => {
                     if key.is_some() {
-                        debug!("Field key duplicated in X-Matrix Authorization header");
+                        debug!("key parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        key = Some(OwnedServerSigningKeyId::try_from(value.to_unescaped())?);
                     }
-                    key = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "sig" => {
                     if sig.is_some() {
-                        debug!("Field sig duplicated in X-Matrix Authorization header");
+                        debug!("key parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        sig = Some(value.to_unescaped());
                     }
-                    sig = Some(std::str::from_utf8(&value).ok()?.to_owned());
                 }
```

If we find a duplicate key, we should throw an error

https://datatracker.ietf.org/doc/html/rfc7235#section-2.1
> Authentication parameters are name=value pairs, where the name token
   is matched case-insensitively, and each parameter name MUST only
   occur once per challenge.

---

Oh yeah, we should also match them case-insensitively.

---

Done.